experiment/migrate_testgrid_tabs.py

```diff
@@ -0,0 +1,197 @@
+#!/usr/bin/env python3
+
+# Copyright 2019 The Kubernetes Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Migrates information from Testgrid's Config.yaml to a subdirectory of Prow Jobs
+
+Only moves Dashboard Tabs
+Skips any Dashboard Tab Configuration that contains unusual keys, even if they're incorrect keys.
+"""
+
+import re
+import argparse
+from os import walk
+import ruamel.yaml
+
+# Prow files that will be ignored
+PROW_BLACKLIST = [
+    # Ruamel won't be able to successfully dump fejta-bot-periodics
+    # See https://bitbucket.org/ruamel/yaml/issues/258/applying-json-patch-breaks-comment
+    "fejta-bot-periodics.yaml",
+    # Generated security jobs are generated with the same name as kubernetes/kubernetes
+    # presubmits, but we never want to migrate to the generated ones.
+    "generated-security-jobs.yaml",
+]
+
+def main(testgrid_config, prow_dir):
+    with open(testgrid_config, "r") as confFile:
+        config = ruamel.yaml.load(confFile,
+                                  Loader=ruamel.yaml.RoundTripLoader,
+                                  preserve_quotes=True)
+
+    for dashboard in config["dashboards"]:
+        if "dashboard_tab" in dashboard:
+            for dashTab in dashboard["dashboard_tab"][:]:
+                if assert_tab_keys(dashTab):
+                    move_tab(dashboard, dashTab, prow_dir)
+
+    with open(testgrid_config, "w") as confFile:
+        ruamel.yaml.dump(config, confFile, Dumper=ruamel.yaml.RoundTripDumper)
+        confFile.truncate()
+
+
+def move_tab(dashboard, dashTab, prow_dir):
+    """ moveTab determines if the dashboard tab matches a prow job in prow_dir.
+    If it does, it modifies the prow job file to include the necessary annotations
+    It also deletes from the config the dashboard tab, so there are no duplicates.
+    """
+    dashboardName = dashboard["name"]
+    prowJobName = dashTab["test_group_name"]
+    prowJobFileName = find_prow_job(prowJobName, prow_dir)
+    if prowJobFileName == "":
+        return
+
+    # Matching file found; patch and delete
+    print("Patching {0} in {1}".format(prowJobName, prowJobFileName))
+
+    with open(prowJobFileName, "r") as prowJobFile:
+        prowConfig = ruamel.yaml.load(prowJobFile,
+                                      Loader=ruamel.yaml.RoundTripLoader,
+                                      preserve_quotes=True)
+
+    # For each presubmits, postsubmits, periodic:
```

Comment:
Argh the logic isn't quite the same for all of them. Could combine presubmits and postsubmits logic at least?

Reply:
I'm not sure that combining these two will increase readability, but cleaning up the loops with the suggestions you've made certainly have